dart: Generate smaller code

brettkail-wk opened the pull request:
### Story:
Use `FGeneratedArgsResultBase` added in #2209 to reduce the size of generated dart code

### Acceptance Criteria:
- [x] Code has been tested and results documented
- [x] Unit tests have been updated
- ~~[ ] Updates to documentation if necessary~~
- ~~[ ] Verify and document changes to any other Messaging components~~
- [x] Pull request made against the 'develop' branch, not master

### Design Notes:
Commits are intended to be reviewed independently.  See commit messages for details.  Ultimately, the diff is +255/-177 on the generator code, and the rest of the changes are to the `compiler/testdata/expected`.

In particular, the removal of the write/read commits sufficiently confuses the diff algorithm to make the test changes unreadable (the `_args`/`_result` class lines no longer delineate the changes).  It's also easier to see the impact of each individual change by reviewing commit-by-commit.

### How To Test:

### My Test Results:
I ran this branch against the full test suite of a large internal application that currently generates ~270KLOC of Dart, and this branch reduces it to ~205KLOC, which reduces the dart2js output by ~340KB.

#### Reviewers:
@Workiva/service-platform
@robbecker-wf @kevinsookocheff-wf

aviary2-wf commented:
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

brettkail-wk commented:
Rebased to resolve two issues:
1. Replaced commit "dart: Fix default values for null_for_unset" with commit "dart: Add TODO to field initialization" commit.  I mistakenly thought that default values weren't being initialized (from code inspection, not testing), but the fields were being initialized when declared, so the old commit was re-initializing fields (so all testing passed), which was already the case for non-`use_null_for_unset`.  Add a comment to make the suboptimal generated code clearer.
4. Added commit "dart: Fix use_null_for_unset default requiredness".  While inspecting the changes of the "dart: Don't generate isSet/unset for _args/_result", I noticed that `isSet` calls were being added to `write`, and that's because `use_null_for_unset` was wrongly checking `isSet` for fields of "primitive" type that were using default requiredness.

adamdockter-wk commented on compiler/generator/base.go:
```diff
@@ -80,8 +80,13 @@ func (b *BaseGenerator) GetElem() string {
-	structs := []*parser.Struct{}
+type Struct struct {
+	*parser.Struct
+	Args bool
```
As discussed in the code review, it would be nice to have a comment next to this boolean to describe what Args == false is since it is not apparent it would mean result struct.

brettkail-wk replied:
Renamed to `ServiceMethodStruct`, and added struct- and field-level doc.

danielszopa-wk commented on compiler/testdata/expected/dart.nullunset/actual_base/f_thing.dart:
```diff
@@ -48,19 +48,11 @@ class thing implements thrift.TBase {
       case AN_ID:
-        if (value == null) {
-          unsetAn_id();
-        } else {
-          this.an_id = value as int;
-        }
+        this.an_id = value as dynamic;
```
This was the chunk that confused me yesterday.

> ... with Null safety enabled, a variable declared as dynamic can be null, while a variable of type Object cannot be null.

My guess is that when we want to support null saftey we can write this like...

```dart
  setFieldValue(int fieldID, dynamic value) {
    switch (fieldID) {
      case AN_ID:
        this.an_id = value;
```

That being said this looks good for now 👍

brettkail-wk replied:
The issue is that it's [implementing this method](https://github.com/Workiva/thrift-dart/blob/3059908dc516048f8aabb24434db2aff7ad1e773/lib/src/t_base.dart#L36), and I don't think we can change that without breaking everything, so we'll probably need the `as dynamic` forever.

@robbecker-wf I would have expected `setFieldValue(int, Object?)` so that the pre-existing `== null` checks continue to work.  Was that an oversight with the null-safety changes?  Similarly, I assume `dynamic?` is correct for the type for `getFieldValue`, but it's not specified at all; is that the default?

brettkail-wk replied:
Discussed offline with Rob, and he agreed that `setFieldValue(int, dynamic)` would break everything and that `setFieldValue(int, Object?)` is the correct signature for null-safety.  There will need to be some changes in the future to add some kind of `--gen dart:null_safe` option.

danielszopa-wk commented on compiler/testdata/expected/dart.int64/actual_base/f_base_foo_service.dart:
```diff
@@ -191,30 +135,6 @@ class basePing_result extends frugal.FGeneratedArgsResultBase {
-    return basePing_result();
-  }
-
   validate() {
```
I'm not familiar with the full semantics of `validate()`, but it seems like we could maybe remove this for args & results as well?

This does look like it might be a bit more complex as you'd have to remove the `validate()` calls from `read()` and `write()` as well. [Here's an example](https://sourcegraph.plat.workiva.net/github.com/Workiva/frugal/-/blob/examples/dart/gen-dart/v1_music/lib/src/f_track.dart?L407-412) where it seems to be generated with an actual implementation.

brettkail-wk commented:
@Workiva/release-management-p

rmconsole-wf reviewed (approved):
+1 from RM
